sklearn/linear_model/cd_fast.pyx

```diff
@@ -259,89 +381,197 @@ def enet_coordinate_descent(np.ndarray[floating, ndim=1] w,
+                                             floating* X_data,
+                                             int* X_indices,
+                                             int* X_indptr,
+                                             floating[:] X_mean,
+                                             floating* y,
+                                             floating y_sum,
+                                             floating[:] R,
+                                             floating R_shift,
+                                             floating[:] w,
+                                             floating[:] XtA,
+                                             floating[:] X_T_R,
+                                             floating[:] dual_scaling,
+                                             floating alpha,
+                                             floating beta,
+                                             bint positive,
+                                             bint center,
+                                             int[:] disabled,
+                                             unsigned int n_disabled=0
+                                             ) nogil:
+
+    cdef floating R_norm2
+    cdef floating R_sum = 0.
+    cdef floating w_norm2
+    cdef floating y_norm2
+    cdef floating l1_norm
+    cdef floating const
+    cdef floating gap
+    cdef unsigned int ii
+    cdef unsigned int jj
+    cdef floating yTA
+    cdef floating dual_norm_XtA
+
+    # fused types version of BLAS functions
+    cdef DOT dot
+    cdef ASUM asum
+
+    if floating is float:
+        dot = sdot
+        asum = sasum
+    else:
+        dot = ddot
+        asum = dasum
+
+    # sparse X.T * R product for non-disabled features
+    if center:
+        # Rtot_sum = R_sum - n_samples * R_shift is constant when
+        # columns are centered
+        R_sum = y_sum + n_samples * R_shift
+
+    for ii in range(n_features):
+        if (n_disabled == 0 or disabled[ii] == 0):
+            X_T_R[ii] = 0.0
+            for jj in range(X_indptr[ii], X_indptr[ii + 1]):
+                X_T_R[ii] += X_data[jj] * R[X_indices[jj]]
+
+            if center:
+                X_T_R[ii] -= X_mean[ii] * R_sum
+            XtA[ii] = X_T_R[ii] - beta * w[ii]
+        else:
+            XtA[ii] = 0.
+
+    if positive:
+        dual_norm_XtA = max(n_features, &XtA[0])
+    else:
+        dual_norm_XtA = abs_max(n_features, &XtA[0])
+
+    yTA = dot(n_samples, &y[0], 1, &R[0], 1)
+
+    # R_norm2 = np.dot(R, R)
+    R_norm2 = dot(n_samples, &R[0], 1, &R[0], 1)
+
+    if center:
+        yTA -= y_sum * R_shift
+        R_norm2 += n_samples * R_shift**2 - 2 * R_sum * R_shift
+
+    if alpha == 0:
+        dual_scaling[0] = 1.
+    else:
+        if dual_norm_XtA <= 0:
+            if R_norm2 == 0:
+                dual_scaling[0] = 1. / alpha
+            else:
+                dual_scaling[0] = yTA / R_norm2 / alpha
```

Comment:
I never really like "a / b / c ". I do tend to prefer "a / (b * c)", which I find more explicit.